dts: bindings: interrupt-controller: GIC: Allow specifying version in DT

p-woj opened the pull request:
This PR separates the GIC binding into 3 version-specific ones and moves the GIC version from the SoCs' Kconfigs into their device trees. This brings the GIC driver more in line with other existing drivers and makes the device tree a more complete description of the hardware, which is useful for efforts relying on structured data such as [Renodepedia](https://zephyr-dashboard.renode.io/renodepedia/).

p-woj committed:
soc: arm64: rk3399: Move GIC version to DT

Move the GIC version to the device tree for rk3399
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm64: am6x: Move GIC version to DT

Move the GIC version to the device tree for am6x
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
dts: bindings: interrupt-controller: GIC: Allow specifying version in DT

Currently, only the presence of a GIC is reflected in the device tree,
and its version must be set separately in each SoC's Kconfig.
This patch adds separate bindings for each GIC version whose presence
in the device tree automatically enables the corresponding Kconfig symbol.

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm: xilinx_zynq7000: Move GIC version to DT

Move the GIC version to the device tree for xilinx_zynq7000
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm64: intel_socfpga: Move GIC version to DT

Move the GIC version to the device tree for intel_socfpga
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm: nxp_s32: Move GIC version to DT

Move the GIC version to the device tree for nxp_s32
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm64: fvp_aemv8: Move GIC version to DT

Move the GIC version to the device tree for fvp_aemv8{a,r}
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm64: viper: Move GIC version to DT

Move the GIC version to the device tree for viper
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm64: mimx9: Move GIC version to DT

Move the GIC version to the device tree for mimx9
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm64: qemu: Move GIC version to DT

Move the GIC version to the device tree for the QEMU platforms
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm64: xenvm: Move GIC version to DT

Move the GIC version to the device tree for xenvm
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm64: nxp_imx: Move GIC version to DT

Move the GIC version to the device tree for nxp_imx
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm: cyclonev: Move GIC version to DT

Move the GIC version to the device tree for cyclonev
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm: renesas_rcar: Move GIC version to DT

Move the GIC version to the device tree for renesas_rcar
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm: xilinx_zynqmp: Move GIC version to DT

Move the GIC version to the device tree for xilinx_zynqmp
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
soc: arm64: ls1046a: Move GIC version to DT

Move the GIC version to the device tree for ls1046a
to improve readability

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
dts: bindings: interrupt-controller: GIC: Update description

When the GIC driver was originally introduced, it was only used on
Cortex-R SoCs. However, this is not the case anymore. Update the
description to reflect that this driver is not specific to Cortex-R.

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

povergoing commented:
LGTM, but someone who has downstream board patches have to update their DTS :)

stephanosio reviewed (commented):
This looks reasonable; but, these changes need to be described in the release notes so that any downstream users using out-of-tree boards can handle the migration.

carlocaione commented on drivers/interrupt_controller/Kconfig.gic:
```diff
@@ -9,21 +9,21 @@ config GIC
 	bool
 
 config GIC_V1
-	bool
+	def_bool DT_HAS_ARM_GIC_V1_ENABLED
```
For consistency this should be:
```
default y
depends on ...
```

p-woj commented:
Thanks. I've added an explanation to the v3.4 release notes draft and converted the Kconfig def_bool to a dependency.

I now see an issue here though, and I'm going to need some advice on how to proceeed. Previously with the `def_bool` approach we (and indeed out-of-tree boards as well) could still use the old method of specifying only `arm,gic` as a compatible string for the GIC and setting its architecture version in Kconfig.

With the `default y` + dependency approach this is not the case anymore - if the GIC version is not present in the DT and the SoC's Kconfig selects it, the build will fail.

This is important not only for out-of-tree boards but also for Zephyr itself because it means that, in order to ensure a buildable state after each individual commit, we would need to first add the new compats to our DTs, then introduce the `DT_HAS` dependency, and only then remove the `select GIC_Vx`s from our Kconfigs. In that case, it might be beneficial to squash some of these commits together, as there are already quite a few considering the simplicity of the change, and this would almost double their number.

Looking forward to your suggestions.

carlocaione commented:
> I now see an issue here though, and I'm going to need some advice on how to proceeed. Previously with the `def_bool` approach we (and indeed out-of-tree boards as well) could still use the old method of specifying only `arm,gic` as a compatible string for the GIC and setting its architecture version in Kconfig.

> With the default y + dependency approach this is not the case anymore - if the GIC version is not present in the DT and the SoC's Kconfig selects it, the build will fail.

This is interesting indeed. @mbolivar @galak do we have a process in place to deprecate a DT compatible?

lorc reviewed (dismissed):
Finally someone did this! Thanks, it is much better in this way.

LGTM for Xen-related changes.

carlocaione reviewed (changes requested):
LGTM but this is blocked on https://github.com/zephyrproject-rtos/zephyr/pull/58035#issuecomment-1554712792 for now

mbolivar-nordic commented:
> do we have a process in place to deprecate a DT compatible?

No formal process, no. From https://github.com/zephyrproject-rtos/zephyr/pull/58035#issuecomment-1554712792, it's just a question of whether to squash some commits together to preserve bisectability, right? That seems reasonable to me.

carlocaione commented:
> No formal process, no. From [#58035 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/58035#issuecomment-1554712792), it's just a question of whether to squash some commits together to preserve bisectability, right? That seems reasonable to me.

No, it is also a matter of breaking out-of-tree ARM64 boards, not just a bisectability issue.

p-woj commented:
I've adjusted the Kconfig changes so that they don't break out-of-tree boards anymore. Now both ways of specifying the architecture version (`select GIC_V2` in Kconfig and `arm,gic-v2` in the DT) will work.

Also, having the version explicitly selected in Kconfig without the versioned compat in the DT prints a deprecation warning.

Please take a look @carlocaione

carlocaione commented on doc/releases/release-notes-3.4.rst:
```diff
@@ -179,6 +179,12 @@ Deprecated in this release
   functionality, enable
   :kconfig:option:`CONFIG_NRF_STORE_REBOOT_TYPE_GPREGRET`.
 
+* Setting the GIC architecture version by selecting
+  :kconfig:option:`CONFIG_GIC_V1`, :kconfig:option:`CONFIG_GIC_V2` and
+  :kconfig:option:`CONFIG_GIC_V3` directly in Kconfig has been deprecated.
+  The GIC version should now be specified by adding the appropriate compatible, for
+  example :dtcompatible:`arm,gic-v2`, to the GIC node in the device tree.
```
This should go now in the 3.5 release notes.

p-woj replied:
Should be good now

carlocaione reviewed (commented):
LGTM. Just a note about the changelog

p-woj committed:
doc: release: 3.5: Add working release notes file

Add a base working release notes file for the Zephyr 3.5 release.

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>

p-woj committed:
doc: release: 3.5: Add note on GIC version selection in DT

Add a note that explains that the GIC version is now specified
in the DT and not Kconfig.

Signed-off-by: Piotr Wojnarowski <pwojnarowski@antmicro.com>
